newrelic/terraform-provider-newrelic
# fix(newrelic_alert_policy_channel): add XAccountID to context when creating or updating alert channels

# Description

This work enables users who have a parent API key to create and link alert channels with their child accounts. 
**This work depends on a PR from newrelic-client-go: https://github.com/newrelic/newrelic-client-go/pull/854**

Fixes https://github.com/newrelic/terraform-provider-newrelic/issues/1645 

## Type of change

Please delete options that are not relevant.


## Checklist:

Please delete options that are not relevant.


## How to test this change?

You will need a parent account API key and a child account to validate these changes work as intended. Please contact me   directly if you experience trouble setting this up.

1) Check out this branch of the newrelic-client-go locally  https://github.com/newrelic/newrelic-client-go/pull/854
2) In the `go.mod` of this provider, replace `newrelic-client-go` with the locally checked out version `replace github.com/newrelic/newrelic-client-go => /path/to/your/local/newrelic-client-go`
3) Build version `99.0.0` of the `terraform-provider-newrelic`
4) Create resources with terraform. You will need to create an alert policy, condition, channel, and associate the channel with the policy.

```
// versions.tf
terraform {
 required_providers {
   newrelic = {
     version = "~> 99.0.0"
     source = "newrelic/newrelic"
   }
 }
 required_version = ">= 0.13"
}
```

```
//main.tf

variable "nr_api_key" {
  type = string
  description = "your new relic API key"
}

variable "account_id" {
  type = number
  description = "your new relic PARENT account"
}

variable "child_account_id" {
  type = number
  description = "your new relic child account"
}

variable "region" {
  type = string
  description = "the region you're using'"
}

provider newrelic {
  api_key = var.nr_api_key
  account_id = var.account_id
  region = var.region
}

resource "newrelic_alert_policy" "my_terraform_test_policy" {
  name = "Child Account Policy"
  account_id = var.child_account_id
}

resource "newrelic_nrql_alert_condition" "my_terraform_test_condition" {
  name = "Child Account Condition"
  policy_id = newrelic_alert_policy.my_terraform_test_policy.id
  nrql {
    query = "SELECT count(*) FROM Transaction"
  }
  critical {
    operator = "above"
    threshold = 0
    threshold_duration = 60
    threshold_occurrences = "all"
  }
  aggregation_window=60
  slide_by=30
  close_violations_on_expiration = true
  value_function = "single_value"
  violation_time_limit_seconds = 86400
  aggregation_method = "event_flow"
  aggregation_delay = 120
  type = "static"
  expiration_duration = 180
  account_id = var.child_account_id
}

resource "newrelic_alert_channel" "my_terraform_test_alert_channel" {
  account_id = var.child_account_id
  name = "Child Account Channel"
  type = "email"
  config {
    recipients              = "foo@bar.com"
    include_json_attachment = "1"
  }
}
# Applies the created channels above to the alert policy
# referenced at the top of the config.
resource "newrelic_alert_policy_channel" "my_terraform_test_alert_policy_channel" {
  policy_id  = newrelic_alert_policy.my_terraform_test_policy.id
  channel_ids = [
    newrelic_alert_channel.my_terraform_test_alert_channel.id,
  ]
  account_id = var.child_account_id
}

```

5) Apply this terraform. In the UI, check the the Policy gets created in the child account, the condition gets created in the child account, the channel gets created in the child account, and the channel is linked to the policy in the child account.
